### PR TITLE
MM-18463 Added update_badge notification when marking a channel as unread

### PR DESCRIPTION
**Pull request by hmhealey**
The most important part of this PR is to have the Android app ignore update_badge notifications so that it doesn't show an empty notification whenever one of those are received. The rest is just adding some constants and cleaning up some things that were bugging me when staring at the Java code trying to make it work.

More information about this is available here: https://github.com/mattermost/mattermost-server/pull/12499

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-18463

### CODE REVIEW

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java [hunk 1]
```diff
@@ -44,8 +45,12 @@
     public static final String KEY_TEXT_REPLY = "CAN_REPLY";
     public static final String NOTIFICATION_REPLIED_EVENT_NAME = "notificationReplied";
 
-    private static LinkedHashMap<String,Integer> channelIdToNotificationCount = new LinkedHashMap<String,Integer>();
-    private static LinkedHashMap<String,List<Bundle>> channelIdToNotification = new LinkedHashMap<String,List<Bundle>>();
+    private static final String PUSH_TYPE_MESSAGE = "message";
+    private static final String PUSH_TYPE_CLEAR = "clear";
+    private static final String PUSH_TYPE_UPDATE_BADGE = "update_badge";
+
+    private static Map<String, Integer> channelIdToNotificationCount = new HashMap<String, Integer>();
```
Comment (enahum) [inline android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java]: What’s the difference between the two and why the change?

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java [hunk 2]
```diff
@@ -131,10 +135,13 @@ public void onReceived() throws InvalidNotificationException {
...
 
-        if ("clear".equals(type)) {
-            cancelNotification(data, notificationId);
-        } else {
+        switch(type) {
+        case PUSH_TYPE_MESSAGE:
             super.postNotification(notificationId);
+            break;
+        case PUSH_TYPE_CLEAR:
+            cancelNotification(data, notificationId);
```
Comment (enahum) [inline android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java]: Dont you need to cancel it for update_badge too?

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java (hunk 1 above)
Comment (hmhealey) [reply android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java]: It's just changing the type of the variables to use the `Map` interface instead of a concrete type. It's just a common practice to do it that way in Java.

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java (hunk 2 above)
Comment (hmhealey) [reply android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java]: The name `cancelNotification` is actually a bit misleading since it's function is to clear an existing notification.

As long as we don't call `postNotification` to create a visible notification, we don't need to do anything to have it ignore the `update_badge` notification. That's the main change here since previously this code would've called `postNotification` when it received an `update_badge`.

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java [hunk 3]
```diff
@@ -105,22 +108,23 @@ public void onReceived() throws InvalidNotificationException {
...
-                list = Collections.synchronizedList(new ArrayList(0));
-            } else {
-                list = Collections.synchronizedList((List)bundleArray);
+            synchronized (channelIdToNotificationCount) {
+                Integer count = channelIdToNotificationCount.get(channelId);
+                if (count == null) {
+                    count = 0;
+                }
+
+                channelIdToNotificationCount.put(channelId, count + 1);
```
Comment (Willyfrog) [inline android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java]: Just for ease of reading it, so people doesn't get confused when value is 1 instead of 0 due to reading too quick.
```suggestion
                     count = 1;
                } else {
                     channelIdToNotificationCount.put(channelId, count + 1);
               }
```

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java [hunk 4]
```diff
@@ -105,22 +108,23 @@ public void onReceived() throws InvalidNotificationException {
...
-            synchronized (list) {
-                if (!"clear".equals(type)) {
+
+            synchronized (channelIdToNotification) {
+                List<Bundle> list = channelIdToNotification.get(channelId);
+                if (list == null) {
+                    list = Collections.synchronizedList(new ArrayList(0));
+                }
+
+                if (PUSH_TYPE_MESSAGE.equals(type)) {
```
Comment (Willyfrog) [inline android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java]: I'm guessing that it's intentional to change from anything but clear, to only message, but can you confirm this change is on purpose? I can only ask, as I'm not aware of the differences between the three types.

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java (hunk 3 above)
Comment (hmhealey) [reply android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java]: I'll move the `count + 1` onto it's own line, but we want to put that value into the map either way, so I'll do it slightly differently than what you've suggested

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java (hunk 4 above)
Comment (hmhealey) [reply android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java]: Yeah, it's intentional. Previously, the two types of notifications were just `message` (show a visible notification for a new message) and `clear` (clear a previous notification). This is adding support for the new `update_badge` notification type that isn't supposed to show a visible notification. `update_badge` would ideally update the badge on the app icon, but that doesn't seem possible on Android, so we're just making Android devices ignore it silently instead of showing a blank notification which is what it would do currently.

This block is adding an entry to `channelIdToNotification[channelId]` for each visible notification, so we only want to do that for `message` notifications. The old logic just had the logic reversed since it was adding an entry for anything that wasn't a `clear`.

**Comment by hmhealey** [comment]
Skipping PR review since this feature will be reviewed as a whole once it's all in
